src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+#include <cstring>
+#include <dlfcn.h>
+
+#include "hooks.h"
+#include "linker_shenanigans.h"
+#include "logging.h"
+#include "macho_utils.h"
+
+#ifndef __APPLE__
+#    error "This file can only be compiled in MacOS systems"
+#endif
+
+namespace memray::linker {
+
+template<typename Hook>
+static void
+patch_symbol(
+        const Hook& hook,
+        typename Hook::signature_t intercept,
+        const char* symname,
+        void* addr,
+        vm_size_t section_size,
+        bool restore_original)
+{
+    vm_size_t trunc_address = trunc_page((vm_size_t)addr);
+    vm_size_t trunc_size = (vm_size_t)addr - trunc_address;
+
+    kern_return_t err = vm_protect(
+            mach_task_self(),
+            (uintptr_t)trunc_address,
+            section_size + trunc_size,
```

Comment:
Do we need to round down to the page boundary? https://web.mit.edu/darwin/src/modules/xnu/osfmk/man/vm_protect.html says

> The region starts at the beginning of the virtual page containing address; it ends at the end of the virtual page containing address + size - 1.

If that's right, it seems like just using `addr` as the address and `section_size` as the size ought to work.

Reply:
Ok, done in the last rebase.